### PR TITLE
Fastah API datasource: initial implementation

### PR DESCRIPTION
**Pull request by s8mathur**
As discussed in issue #81 . Review welcome, and appreciated. 
Also, it's a bit daunting to create the test setup on my local sandbox, so I will need to see the Travis bot output.

### CODE REVIEW

**Comment by benjaminpick** [comment]
Exciting! I will add a few comments to you code ...

**Comment by benjaminpick** [comment]
I will merge it into a feature branch and do the small changes above myself, you can continue to improve it there if you want. I would release it for 5.1.0 in some week's time.

File: data-sources/fastah.php [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+            // Continent
+            $r['continent']['code'] = strtoupper($data['locationData']['continentCode']);
+            // Country
+            $r['country']['names'] = $this->locales($locale, $data['locationData']['countryName']);
+            $r['country']['iso_code'] = strtoupper($data['locationData']['countryCode']);
+            // City
+            $r['city']['names'] = $this->locales($locale, $data['locationData']['cityName']);
+            $r['city']['geoname_id'] = $data['locationData']['cityGeonamesId'];
+            // Lat, Lng
+            $r['location']['latitude'] = $data['locationData']['lat'];
```
[the remark this answers is not in the document]
Comment (benjaminpick) [reply data-sources/fastah.php:96]: I have added !empty / isset now

File: views/options.php [hunk 2]
```diff
@@ -85,6 +85,9 @@
 			<?php if (in_array($currentSourceId, array('precision', 'ipstack')) && !empty($wp_options['ajax_enabled'])): ?>
 				<span class="geoip_detect_error" style="margin-top: 0;"><?php printf(__('Warning: In theory, other websites could use your API credits over AJAX, this cannot be prevented completely (see <a href="%s" target="_blank">documentation</a> for more infos). You should use a different data source or disable AJAX.', 'geoip-detect'), 'https://github.com/yellowtree/geoip-detect/wiki/JS-API-Documentation'); ?></span>
 			<?php endif; ?>
+			<?php if (in_array($currentSourceId, array('precision', 'fastah')) && !empty($wp_options['ajax_enabled'])): ?>
```
[the remark this answers is not in the document]
Comment (benjaminpick) [reply views/options.php:88]: Removing this extra condition

**Comment by benjaminpick** [comment]
The lookup is working now, but I am getting response times of 700-800ms with HTTP/2 enabled. Do you think it is because it is currently using the Wordpress implementation `wp_remote_get` ? I tried stream_context but it wasn't faster. Maybe we do have to implent curl_init manually?

**Comment by benjaminpick** [comment]
BTW, the plugin already does client caching of the response, both in-memory and in-database (across lookups). 
Most of the times, it won't be used for doing many consequent lookups of different IP adresses (because most use cases just need the IP adress of the current client), so optimizing for concurrency wouldn't speed it up here either. Hm ... Please try it out on your dev box and let me know your timings, both for servers on aws and outside (I guess your prospective clients also will be either inside or outside the AWS world, right?)

**Comment by benjaminpick** [comment]
https://console.api.getfastah.com is also quite slow for me, with response times of 1-5 seconds for a page load. Other sites are fast, maybe your AWS datacenter is too far away from Germany?

**Comment by s8mathur** [comment]
> BTW, the plugin already does client caching of the response, both in-memory and in-database (across lookups).
> Most of the times, it won't be used for doing many consequent lookups of different IP adresses (because most use cases just need the IP adress of the current client), so optimizing for concurrency wouldn't speed it up here either. Hm ... Please try it out on your dev box and let me know your timings, both for servers on aws and outside (I guess your prospective clients also will be either inside or outside the AWS world, right?)

In the interest of speed of getting version 1 available to plugin users, let's remove the HTTP/2 support entirely and use the caching that you refer to as the latency-minimizer. 

PS: As seen on these 2 pages, the HTTP/2 multiplexing feature needs some testing to figure out, including async use of the multi object: 
* https://curl.se/docs/http2.html 
* https://www.php.net/manual/en/function.curl-multi-init.php

**Pull request merged**